pkg/state/state_doca_telemetry_service.go

```diff
@@ -0,0 +1,196 @@
+	reqLogger.WithValues("State:", d.name, "Name:", cr.Name, "Namespace:", cr.Namespace)
+	log.IntoContext(ctx, reqLogger)
+	reqLogger.V(consts.LogLevelInfo).Info("Sync Custom resource")
+
+	if cr.Spec.DOCATelemetryService == nil {
+		// Either this state was not required to run or an update occurred and we need to remove
+		// the resources that where created.
+		return d.handleStateObjectsDeletion(ctx)
+	}
+
+	objs, err := d.GetManifestObjects(ctx, cr, infoCatalog, reqLogger)
+	if err != nil {
+		return SyncStateNotReady, errors.Wrap(err, "failed to create k8s objects from manifest")
+	}
+	if len(objs) == 0 {
+		return SyncStateNotReady, nil
+	}
+
+	// Create objects if they don't exist, Update objects if they do exist
+	err = d.createOrUpdateObjs(ctx, func(obj *unstructured.Unstructured) error {
+		if err := controllerutil.SetControllerReference(cr, obj, d.client.Scheme()); err != nil {
+			return errors.Wrap(err, "failed to set controller reference for object")
+		}
+		return nil
+	}, objs)
+	if err != nil {
+		return SyncStateNotReady, errors.Wrap(err, "failed to create/update objects")
+	}
+	waitForStaleObjectsRemoval, err := d.handleStaleStateObjects(ctx, objs)
+	if err != nil {
+		return SyncStateNotReady, errors.Wrap(err, "failed to handle state stale objects")
+	}
+	if waitForStaleObjectsRemoval {
+		return SyncStateNotReady, nil
+	}
+	// Check objects status
+	syncState, err := d.getSyncState(ctx, objs)
+	if err != nil {
+		return SyncStateNotReady, errors.Wrap(err, "failed to get sync state")
+	}
+	return syncState, nil
+}
+
+// GetManifestObjects returns the Unstructured objects to deploy for DOCA Telemetry Service.
+func (d docaTelemetryServiceState) GetManifestObjects(
+	_ context.Context, cr *mellanoxv1alpha1.NicClusterPolicy,
+	_ InfoCatalog, reqLogger logr.Logger) ([]*unstructured.Unstructured, error) {
+	if cr == nil || cr.Spec.DOCATelemetryService == nil {
+		return nil, errors.New("failed to render objects: state spec is nil")
+	}
+	dts := cr.Spec.DOCATelemetryService
+
+	configMapName := docaTelemetryServiceDefaultConfigMapName
+	if dts.Config != nil {
+		configMapName = dts.Config.FromConfigMap
+	}
+	renderData := &DOCATelemetryServiceManifestRenderData{
+		CrSpec:        dts,
+		ConfigMapName: configMapName,
+		Tolerations:   cr.Spec.Tolerations,
+		NodeAffinity:  cr.Spec.NodeAffinity,
+		RuntimeSpec: &dtsRuntimeSpec{
+			runtimeSpec:        runtimeSpec{config.FromEnv().State.NetworkOperatorResourceNamespace},
+			ContainerResources: createContainerResourcesMap(cr.Spec.DOCATelemetryService.ContainerResources),
+		},
+	}
+
+	// Render objects related to the DOCATelemetryService
+	reqLogger.V(consts.LogLevelDebug).Info("Rendering objects", "data:", renderData)
+	renderedObjects, err := d.renderer.RenderObjects(&render.TemplatingData{Data: renderData})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to render objects")
+	}
+
+	// Remove the configMap from the deployment if an explicit config is set.
```

Comment:
We can - is it preferable? 

My preference is to do any logic that's unique to this state in go code because debugging templates is annoying - but I can move this logic to the template if we want.

Reply:
Better to do it in template.
We have already existing logic that will remove the stale objects based on the state label

Reply:
That logic is being used here - this code doesn't delete the object - it just removes it from the list of desired objects.

I can definitely move this to the template file though